File: my-folder/0021-merge-two-sorted-lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:

    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        #Check if any one of the list is none->Return not none
        #Start iterating while checking which one is smaller and go on forming the linked list
        if(list1 is None):
            return list2
        if(list2 is None):
            return list1
        
        h1=list1
        h2=list2


        finalHead=None
        finalTail=None
        
        while(h1 is not None and h2 is not None):
            if(h1.val<=h2.val):
                if(finalHead is None):
                    finalHead=h1
                    finalTail=h1
                    h1=h1.next
                else :
                    #move final tail to h1
                    #add h1 to the final tail
                    #move final tail to the last node
                    #move h1 one step
                    finalTail.next=h1
                    finalTail=h1
                    h1=h1.next

            else:
                if(finalHead is None):
                    finalHead=h2
                    finalTail=h2
                    h2=h2.next
                else :
                    #move final tail to h1
                    #add h1 to the final tail
                    #move final tail to the last node
                    #move h1 one step
                    finalTail.next=h2
                    finalTail=h2
                    h2=h2.next
        if(h1 is None):
            finalTail.next=h2
        if(h2 is None):
            finalTail.next=h1

        return finalHead
                    

        

File: my-folder/0021-merge-two-sorted-lists/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        //I want to do a recursive solution for this 
        //Compare the current two node and return whichever is the smaller
        if(list1==null){
            return list2;
        }
        if(list2==null){
            return list1;
        }

        ListNode mergedNode;

        if(list1.val<list2.val){
            list1.next=mergeTwoLists(list1.next,list2);
            mergedNode=list1;
        }
        else{
            list2.next=mergeTwoLists(list1,list2.next);
            mergedNode=list2;

        }

        return mergedNode;
    }
}


// /**
//  * Definition for singly-linked list.
//  * public class ListNode {
//  *     int val;
//  *     ListNode next;
//  *     ListNode() {}
//  *     ListNode(int val) { this.val = val; }
//  *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
//  * }
//  */
// class Solution {
//     public ListNode mergeTwoLists(ListNode list1, ListNode list2) {

//         //Compare the heads of both the list and make a new list
//         //Dummy node to track the head of the merged list easily
//         ListNode dummy=new ListNode(-1);

//         ListNode current =dummy;

//         while(list1!=null && list2!=null){
//             if(list1.val<list2.val){
//                 current.next=list1;
//                 current=current.next;
//                 list1=list1.next;
//             }
//             else{
//                 current.next=list2;
//                 current=current.next;
//                 list2=list2.next;
//             }             
//         }

//         if(list1==null){
//             current.next=list2;
//         }
//         else{
//             current.next=list1;
//         }

//         return dummy.next;

//     }
// }




